# HG changeset patch
# User Rik <rik@octave.org>
# Date 1472498308 25200
#      Mon Aug 29 12:18:28 2016 -0700
# Node ID cc3be522ec79be30ed0bb39da8ce4ef06d48c84a
# Parent  4caa7b28d18347de79a4433e21649670e6f44339
rose.m: Fix plotting that was off by 180 degrees (bug #48889).

* rose.m: Correctly normalize input theta to range [0, 2*pi].
Add check for bin sizes >= pi.  Report a warning if these are found.
Add input validation BIST tests.

diff --git a/scripts/plot/draw/rose.m b/scripts/plot/draw/rose.m
--- a/scripts/plot/draw/rose.m
+++ b/scripts/plot/draw/rose.m
@@ -56,40 +56,50 @@
 function [thout, rout] = rose (varargin)
 
   [hax, varargin, nargin] = __plt_get_axis_arg__ ("rose", varargin{:});
 
   if (nargin < 1)
     print_usage ();
   endif
 
+  th = varargin{1};
   ## Force theta to [0,2*pi] range
-  th = varargin{1};
-  th = atan2 (sin (th), cos (th)) + pi;
+  th = atan2 (sin (th), cos (th));
+  th(th < 0) += 2*pi;
 
+  do_bin_check = false;
   if (nargin > 1)
     x = varargin{2};
     if (isscalar (x))
+      do_bin_check = (x < 3); 
       x = [0.5/x : 1/x : 1] * 2*pi;
     else
       ## Force theta to [0,2*pi] range
-      x = atan2 (sin (x), cos (x)) + pi;
+      x = atan2 (sin (x), cos (x));
+      x(x < 0) += 2*pi;
+      x = sort (x);
+      do_bin_check = true;
     endif
   else
     x = [1/40 : 1/20 : 1] * 2*pi;
   endif
 
   [nn, xx] = hist (th, x);
   xx = xx(:).';
   if (isvector (nn))
     nn = nn(:);
   endif
   x1 = xx(1:end-1) + diff (xx, 1) / 2;
   x1 = [x1 ; x1; x1; x1](:);
   th = [0; 0; x1; 2*pi ; 2*pi];
+  if (do_bin_check && any (diff (th) >= pi))
+    warning ("rose: bin sizes >= pi will not plot correctly");
+  endif
+
   r = zeros (4 * rows (nn), columns (nn));
   r(2:4:end, :) = nn;
   r(3:4:end, :) = nn;
 
   if (nargout < 2)
     oldfig = [];
     if (! isempty (hax))
       oldfig = get (0, "currentfigure");
@@ -119,8 +129,13 @@ endfunction
 %! rose (2*randn (1e5, 1), 8);
 %! title ("rose() angular histogram plot with 8 bins");
 
 %!demo
 %! clf;
 %! rose ([2*randn(1e5, 1), pi + 2*randn(1e5, 1)]);
 %! title ("rose() angular histogram plot with 2 data series");
 
+## Test input validation
+%!error rose ()
+%!warning <bin sizes .= pi will not plot correctly>
+%! [th, r] = rose ([1 2 2 4 4 4], [1 2 pi]);
+
